.github/workflows/update-versions.yml

```diff
@@ -0,0 +1,198 @@
+      run: |
+        TAG_COMMIT=$(git rev-list -n 1 ${{ steps.get_version.outputs.VERSION }})
+        echo "Tag points to commit: $TAG_COMMIT"
+
+        if ! git merge-base --is-ancestor $TAG_COMMIT origin/main; then
+          echo "❌ Error: Tag ${{ steps.get_version.outputs.VERSION }} does not originate from main branch"
+          exit 1
+        fi
+
+        echo "✅ Tag ${{ steps.get_version.outputs.VERSION }} originates from main branch"
+
+  build-and-push-run-image:
+    runs-on: ubuntu-latest
+    needs: validate-tag
+    permissions:
+      contents: read
+      packages: write # Grant write permission for pushing to GHCR
+
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v4
+
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v3
+      with:
+        driver: docker-container
+
+    - name: Log in to GitHub Container Registry
+      uses: docker/login-action@v3
+      with:
+        registry: ghcr.io
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Build and push run-image
+      uses: docker/build-push-action@v5
+      with:
+        context: ./run-image
+        file: ./run-image/Dockerfile
+        push: true
+        tags: ghcr.io/${{ needs.validate-tag.outputs.repo_lower }}/base-image:${{ needs.validate-tag.outputs.version }}
+        cache-from: type=gha
+        cache-to: type=gha,mode=max
+
+  update-versions:
+    runs-on: ubuntu-latest
+    needs: validate-tag
+    permissions:
+      contents: write # Grant write permission to the GITHUB_TOKEN
+
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v4
+      with:
+        fetch-depth: 0
+
+    - name: Switch to main branch
+      run: |
+        git checkout main
+        git pull origin main
+
+    - name: Update buildpack.toml versions
+      run: |
+        RELEASE_VERSION=${{ needs.validate-tag.outputs.version }}
+        find buildpacks -name "buildpack.toml" | while read file; do
+          sed -i "s|^version = \".*\"|version = \"${RELEASE_VERSION}\"|g" "$file"
+          echo "Updated $file to version $RELEASE_VERSION"
+        done
+
+    - name: Update builders/selector/builder.toml
+      run: |
+        RELEASE_VERSION=${{ needs.validate-tag.outputs.version }}
+        BUILDER_FILE="builders/selector/builder.toml"
+        REPO_LOWER="${{ needs.validate-tag.outputs.repo_lower }}"
+
+        # Create a temporary script to handle TOML section parsing
+        cat > update_builder.sh << 'EOF'
+        #!/bin/bash
+        RELEASE_VERSION="$1"
+        BUILDER_FILE="$2"
+        REPO_LOWER="$3"
+
+        # Use awk to properly handle TOML sections for buildpacks
+        awk -v version="$RELEASE_VERSION" '
+        /^\[\[buildpacks\]\]/ {
+            print
+            in_buildpack=1
+            local_buildpack=0
+            next
+        }
+        /^\[\[/ && !/^\[\[buildpacks\]\]/ {
+            in_buildpack=0
+            local_buildpack=0
+        }
```

Comment:
issue: All of the bash code here should move to a dedicated bash script that can be invoked from the github action.